text/0000-improved-ember-registry-apis.md

```diff
@@ -0,0 +1,719 @@
+
+### TypeScript motivation
+
+The type definitions for Ember currently make use of a pattern we call a “type registry”: a mapping of *strings* to *types*. This allows the types to safely represent string-based lookups like `story.findRecord('some-model')` or `service('session')`.
+
+For example, to make service lookups work with the classic API, users had to write something like this:
+
+```ts
+import Service from '@ember/service';
+
+export default class Session extends Service {
+  login(email: string, password: string) {
+    // ...
+  }
+  
+  logout() {
+    // ...
+  }
+}
+
+declare module '@ember/service' {
+  interface Registry {
+    session: Session;
+  }
+}
+```
+
+Then, users could write fairly idiomatic Ember and have the types resolve correctly:
+
+```ts
+import Component from '@ember/component';
+import { inject as service } from '@ember/service';
+
+export default Component.extend({
+  session: service('session'), // type: `Session`
+})
+```
+
+The Typed Ember blueprints generate that boilerplate for users, which (hopefully) minimizes the pain of working with these definitions. However, to make this work for the registry APIs, users (and therefore the blueprints) would have to add *another* declaration to the bottom of the file:
+
+```diff
+  import Service from '@ember/service';
+  
+  export default class Session extends Service {
+    login(email: string, password: string) {
+      // ...
+    }
+    
+    logout() {
+      // ...
+    }
+  }
+  
+  declare module '@ember/service' {
+    interface Registry {
+      session: Session;
+    }
+  }
+  
++ declare module 'TBD-some-registry-spot' {
++   interface Registry {
++     'service:session': typeof Session;
++   }
++ }
+```
+
+This *works*, but is additional boilerplate and, more importantly, *duplication of information*, which can get out of sync. Both of these problems are avoided by the new design proposed by this RFC. With the new design, users would still need the same boilerplate required today, but the types for `lookup` and other registry APIs can be extended in a way transparent to users so that they “just work” with *zero* additional effort, as covered in the next section.
+
+### Proposed type definitions
+
+If we were to ship this, we would make one significant change to the *existing* types on DefinitelyTyped, supplying a codemod for ember-cli-typescript users. Where today, the various type registries refer to the class types, we would update them to refer to the *constructor* type for the class. For example, for the service registry:
+
+```diff
+ interface Registry {
+-  session: Session;
++  session: typeof Session;
```

Comment:
Do you have any concerns about the global-ness of `Registry` types and how that plays with the isolated registries/containers that engines have? For engines developed in a separate repo from each other/their host app, the possibility of cross-contamination is probably pretty low, but in a monorepo or in-app-engine setup, it seems like projects are likely to wind up with the union of all the available services/controllers/etc.

I realize this is already the status quo, but since this RFC touches on and expands the use of type registries, it seemed like it might be worth raising, even if the answer is ultimately still "it's probably fine."

Reply:
My take is roughly: It's *not* really fine, and we need to develop a solution for it—it's on my roadmap for work I expect to be doing in the last three quarters of this year 🤞—but (a) the ideas introduced here don't meaningfully *worsen* the situation and (b) the solution we eventually build to solve it *should* work just as well.

I have some still-very-unformed ideas around ways we could leverage the changes proposed in this RFC, particularly around the use of namespaces. We might, for example, be able to tweak the registries to include the namespace for addons (including engines), and then (a) recommend to TS users that they consistently use the namespace rather than letting it be inferred or (b) see if we can figure out some way to infer based on context what the namespace should be [probably impossible?] or (c) do something that combines (a) with TS' project configurations to get the right outcomes. Like I said: *very* un-formed as of yet, but the ideas become much more tenable if this RFC ends up accepted and the API is rationalized accordingly.